File: samples/bot-formatting-cards/nodejs/bots/botFormattingCards.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ActivityHandler, MessageFactory, CardFactory } = require('botbuilder');
const { ActionTypes } = require('botframework-schema');
const MentionSupport = require('../resources/mentionSupport.json');
const InformationMaskingCard = require('../resources/informationMasking.json');
const SampleAdaptiveCard = require('../resources/sampleAdaptiveWithFullWidth.json');
const StageViewImagesCard = require('../resources/stageViewForImages.json');
const OverFlowMenuCard = require('../resources/overflowMenu.json');
const HTMLConnectorCard = require('../resources/formatHTMLConnectorCard.json');
const CardWithEmoji = require('../resources/adaptiveCardWithEmoji.json');
const PeoplePersonaCardIcon = require('../resources/adaptivePeoplePersonaCardIcon.json');
const PeoplePersonaCardSetIcon = require('../resources/adaptivePeoplePersonaCardSetIcon.json');
const CodeBlocksCard = require('../resources/codeBlocksCard.json');
const AdaptiveCardResponsiveLayout = require('../resources/AdaptiveCardResponsiveLayout.json');
const AdaptiveCardBorders = require('../resources/adaptiveCardBorders.json');
const AdaptiveCardRoundedCorners = require('../resources/adaptiveCardRoundedCorners.json');
const adaptiveCardFluentIcons = require('../resources/adaptiveCardFluentIcon.json');
const adaptiveCardMediaElements = require('../resources/adaptiveCardMediaElements.json');
const adaptiveCardStarRatings = require('../resources/adaptiveCardStarRatings.json');
const adaptiveCardConditional = require('../resources/adaptiveCardConditional.json');
const adaptiveCardScrollable = require('../resources/adaptiveCardScrollable.json');
const adaptiveCardCompoundButton = require('../resources/adaptiveCardCompoundButton.json');
const adaptiveCardContainerLayouts = require('../resources/adaptiveCardContainerLayouts.json');
const adaptiveCardDonutChart = require('../resources/adaptiveCardDonutChart.json');
const adaptiveCardGaugeChart = require('../resources/adaptiveCardGaugeChart.json');
const adaptiveCardHorizontalBarChart = require('../resources/adaptiveCardHorizontalBarChart.json');
const adaptiveCardHorizontalBarStackedChart = require('../resources/adaptiveCardHorizontalBarStacked.json');
const adaptiveCardLineChart = require('../resources/adaptiveCardLineChart.json');
const adaptiveCardPieChart = require('../resources/adaptiveCardPieChart.json');
const adaptiveCardVerticalBarChart = require('../resources/adaptiveCardVerticalBarChart.json');
const adaptiveCardVerticalBarGroupedChart = require('../resources/adaptiveCardVerticalBarGroupedChart.json');

/**
 * BotFormattingCards class handles the bot's activities and responses.
 */
class BotFormattingCards extends ActivityHandler {
    constructor() {
        super();

        this.onMembersAdded(async (context, next) => {
            await this.sendWelcomeMessage(context);
            await next();
        });

        this.onMessage(async (context, next) => {
            const text = context.activity.text;

            const adaptiveFormatCards = {
                'CodeBlock': this.sendCodeBlock,
                'MentionSupport': this.sendMentionSupportCard,
                'InfoMasking': this.sendInfoMasking,
                'FullWidthCard': this.sendFullWidthCard,
                'StageViewImages': this.sendStageViewImagesCard,
                'OverflowMenu': this.sendOverFlowMenuCard,
                'HTMLConnector': this.sendHTMLConnectorCard,
                'CardWithEmoji': this.sendCardWithEmoji,
                'Persona': this.sendPersonaCardIcons,
                'PersonaSet': this.sendPersonaCardSetIcons,
                'Layout': this.sendLayoutCard,
                'Borders': this.sendBordersCard,
                'RoundedCorners': this.sendRoundedCornersCard,
                'FluentIcons': this.sendFluentIconsCard,
                'MediaElements': this.sendMediaElementsCard,
                'StarRatings': this.sendStarRatingsCard,
                'ConditionalCard': this.sendConditionalCard,
                'ScrollableContainer': this.sendScrollableCard,
                'CompoundButton': this.sendCompoundButtonCard,
                'ContainerLayout': this.sendContainerLayoutCard,
                'DonutChart': this.sendDonutChartCard,
                'GaugeChart': this.sendGaugeChartCard,
                'HorizontalChart': this.sendHorizontalBarChartCard,
                'HorizontalChartStacked': this.sendHorizontalBarStackedChartCard,
                'LineChart': this.sendLineChartCard,
                'PieChart': this.sendPieChartCard,
                'VerticalBarChart': this.sendVerticalBarChartCard,
                'VerticalBarGroupedChart': this.sendVerticalBarGroupedChartCard
            };

            if (adaptiveFormatCards[text]) {
                await context.sendActivity({ attachments: [adaptiveFormatCards[text].call(this)] });
                await context.sendActivity(`You have Selected <b>${text}</b>`);
            } else if (context.activity.value != null && context.activity.text == undefined) {
                const activityValue = context.activity.value;
                if (activityValue.hasOwnProperty('rating1') && activityValue.hasOwnProperty('rating2')) {
                    await context.sendActivity(`Ratings Feedback: ${JSON.stringify(activityValue)}`);
                }
            }

            await this.sendAdaptiveCardFormats(context);
            await next();
        });
    }

    /**
     * Send a welcome message along with Adaptive card format actions for the user to click.
     * @param {TurnContext} turnContext A TurnContext instance containing all the data needed for processing this conversation turn.
     */
    async sendWelcomeMessage(turnContext) {
        const { activity } = turnContext;
        for (const idx in activity.membersAdded) {
            if (activity.membersAdded[idx].id !== activity.recipient.id) {
                const welcomeMessage = `Welcome to Adaptive Card Format. This bot will introduce you to different types of formats. Please select the cards from given options`;
                await turnContext.sendActivity(welcomeMessage);
                await this.sendAdaptiveCardFormats(turnContext);
            }
        }
    }

    /**
     * Sends Mention Support Card
     */
    sendMentionSupportCard() {
        return CardFactory.adaptiveCard(MentionSupport);
    }

    sendCodeBlock() {
        return CardFactory.adaptiveCard(CodeBlocksCard);
    }

    /**
     * Adaptive Card updated to be responsive using targetWidth.
     */
    sendLayoutCard() {
        return CardFactory.adaptiveCard(AdaptiveCardResponsiveLayout);
    }

    /**
     * Sends Sample Adaptive Card With Full Width
     */
    sendFullWidthCard() {
        return CardFactory.adaptiveCard(SampleAdaptiveCard);
    }

    /**
     * Sends StageView Images Card
     */
    sendStageViewImagesCard() {
        return CardFactory.adaptiveCard(StageViewImagesCard);
    }

    /**
     * Sends InfoMasking Card
     */
    sendInfoMasking() {
        return CardFactory.adaptiveCard(InformationMaskingCard);
    }

    /**
     * Sends OverFlow Menu Card
     */
    sendOverFlowMenuCard() {
        return CardFactory.adaptiveCard(OverFlowMenuCard);
    }

    /**
     * Sends HTML Connector Card
     */
    sendHTMLConnectorCard() {
        return CardFactory.o365ConnectorCard(HTMLConnectorCard);
    }

    /**
     * Sends Card With Emoji
     */
    sendCardWithEmoji() {
        return CardFactory.adaptiveCard(CardWithEmoji);
    }

    /**
     * Persona card Icon in an Adaptive Card
     * If you want to show a single user in an Adaptive Card, the Adaptive Card displays the people icon and the name of the user.
     */
    sendPersonaCardIcons() {
        return CardFactory.adaptiveCard(PeoplePersonaCardIcon);
    }

    /**
     * Persona Card Set Icon in an Adaptive Card
     * If you want to show multiple users in an Adaptive Card, the Adaptive Card displays only the people icon of the users.
     */
    sendPersonaCardSetIcons() {
        return CardFactory.adaptiveCard(PeoplePersonaCardSetIcon);
    }

    /**
     * Sends Card showing the use of Borders on columns, columnsets, containers, etc.
     */
    sendBordersCard() {
        return CardFactory.adaptiveCard(AdaptiveCardBorders);
    }

    /**
     * Sends Card showing the use of Rounded Corners on columns, columnsets, containers, tables, etc.
     */
    sendRoundedCornersCard() {
        return CardFactory.adaptiveCard(AdaptiveCardRoundedCorners);
    }

    /**
     * Generates an Adaptive Card attachment that includes Fluent icons.
     * This method reads the adaptive card JSON from a resource file and deserializes it to be included as an attachment with Fluent icons in the card.
     * Fluent icons provide a modern and visually appealing way to enhance the UI within Adaptive Cards.
     */
    sendFluentIconsCard() {
        return CardFactory.adaptiveCard(adaptiveCardFluentIcons);
    }

    /**
     * Creates and returns an Attachment containing an adaptive card with media elements.
     */
    sendMediaElementsCard() {
        return CardFactory.adaptiveCard(adaptiveCardMediaElements);
    }

    /**
     * Sends a star ratings card as an attachment for displaying or collecting user feedback.
     */
    sendStarRatingsCard() {
        return CardFactory.adaptiveCard(adaptiveCardStarRatings);
    }

    /**
     * Sends a Conditional Action.submit button enable card
     */
    sendConditionalCard() {
        return CardFactory.adaptiveCard(adaptiveCardConditional);
    }

    /**
     * Sends a Scrollable container adaptive card
     */
    sendScrollableCard() {
        return CardFactory.adaptiveCard(adaptiveCardScrollable);
    }

    /**
     * Sends a Compound Button adaptive card
     */
    sendCompoundButtonCard() {
        return CardFactory.adaptiveCard(adaptiveCardCompoundButton);
    }

    /**
     * Sends a Container Layout adaptive card
     */
    sendContainerLayoutCard() {
        return CardFactory.adaptiveCard(adaptiveCardContainerLayouts);
    }

    /**
     * Sends a Donut Chart adaptive card
     */
    sendDonutChartCard() {
        return CardFactory.adaptiveCard(adaptiveCardDonutChart);
    }

    /**
     * Sends a Gauge Chart adaptive card
     */
    sendGaugeChartCard() {
        return CardFactory.adaptiveCard(adaptiveCardGaugeChart);
    }

    /**
     * Sends a Horizontal Bar Chart adaptive card
     */
    sendHorizontalBarChartCard() {
        return CardFactory.adaptiveCard(adaptiveCardHorizontalBarChart);
    }

    /**
     * Sends a Horizontal Bar Stacked Chart adaptive card
     */
    sendHorizontalBarStackedChartCard() {
        return CardFactory.adaptiveCard(adaptiveCardHorizontalBarStackedChart);
    }

    /**
     * Sends a Line Chart adaptive card
     */
    sendLineChartCard() {
        return CardFactory.adaptiveCard(adaptiveCardLineChart);
    }

    /**
     * Sends a Pie Chart adaptive card
     */
    sendPieChartCard() {
        return CardFactory.adaptiveCard(adaptiveCardPieChart);
    }

    /**
     * Sends a Vertical Bar Chart adaptive card
     */
    sendVerticalBarChartCard() {
        return CardFactory.adaptiveCard(adaptiveCardVerticalBarChart);
    }

    /**
     * Sends a Vertical Bar Grouped Chart adaptive card
     */
    sendVerticalBarGroupedChartCard() {
        return CardFactory.adaptiveCard(adaptiveCardVerticalBarGroupedChart);
    }

    /**
     * Send AdaptiveCard Formats to the user.
     * @param {TurnContext} turnContext A TurnContext instance containing all the data needed for processing this conversation turn.
     */
    async sendAdaptiveCardFormats(turnContext) {
        const cardActions = [
            { type: ActionTypes.ImBack, title: 'MentionSupport', value: 'MentionSupport' },
            { type: ActionTypes.ImBack, title: 'InfoMasking', value: 'InfoMasking' },
            { type: ActionTypes.ImBack, title: 'FullWidthCard', value: 'FullWidthCard' },
            { type: ActionTypes.ImBack, title: 'StageViewImages', value: 'StageViewImages' },
            { type: ActionTypes.ImBack, title: 'OverflowMenu', value: 'OverflowMenu' },
            { type: ActionTypes.ImBack, title: 'HTMLConnector', value: 'HTMLConnector' },
            { type: ActionTypes.ImBack, title: 'CardWithEmoji', value: 'CardWithEmoji' },
            { type: ActionTypes.ImBack, title: 'Persona', value: 'Persona' },
            { type: ActionTypes.ImBack, title: 'PersonaSet', value: 'PersonaSet' },
            { type: ActionTypes.ImBack, title: 'CodeBlock', value: 'CodeBlock' },
            { type: ActionTypes.ImBack, title: 'Layout', value: 'Layout' },
            { type: ActionTypes.ImBack, title: 'Borders', value: 'Borders' },
            { type: ActionTypes.ImBack, title: 'RoundedCorners', value: 'RoundedCorners' },
            { type: ActionTypes.ImBack, title: 'FluentIcons', value: 'FluentIcons' },
            { type: ActionTypes.ImBack, title: 'MediaElements', value: 'MediaElements' },
            { type: ActionTypes.ImBack, title: 'StarRatings', value: 'StarRatings' },
            { type: ActionTypes.ImBack, title: 'ConditionalCard', value: 'ConditionalCard' },
            { type: ActionTypes.ImBack, title: 'ScrollableContainer', value: 'ScrollableContainer' },
            { type: ActionTypes.ImBack, title: 'CompoundButton', value: 'CompoundButton' },
            { type: ActionTypes.ImBack, title: 'ContainerLayout', value: 'ContainerLayout' },
            { type: ActionTypes.ImBack, title: 'DonutChart', value: 'DonutChart' },
            { type: ActionTypes.ImBack, title: 'GaugeChart', value: 'GaugeChart' },
            { type: ActionTypes.ImBack, title: 'HorizontalChart', value: 'HorizontalChart' },
            { type: ActionTypes.ImBack, title: 'HorizontalChartStacked', value: 'HorizontalChartStacked' },
            { type: ActionTypes.ImBack, title: 'LineChart', value: 'LineChart' },
            { type: ActionTypes.ImBack, title: 'PieChart', value: 'PieChart' },
            { type: ActionTypes.ImBack, title: 'VerticalBarChart', value: 'VerticalBarChart' },
            { type: ActionTypes.ImBack, title: 'VerticalBarGroupedChart', value: 'VerticalBarGroupedChart' }
        ];

        await this.sendWelcomeCard(turnContext, cardActions);
    }

    /**
     * Sends a welcome card with given card actions.
     * @param {TurnContext} context A TurnContext instance containing all the data needed for processing this conversation turn.
     * @param {Array} cardActions Array of card actions to be included in the welcome card.
     */
    async sendWelcomeCard(context, cardActions) {
        const card = CardFactory.heroCard(
            'Please select a card from given options.',
            '',
            null,
            cardActions
        );
        await context.sendActivity(MessageFactory.attachment(card));
    }
}

module.exports.BotFormattingCards = BotFormattingCards;

File: samples/bot-formatting-cards/python/bots/botFormattingCards.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.core.teams import TeamsActivityHandler
from adaptive_cards import cards
from botbuilder.schema.teams import TeamInfo, TeamsChannelAccount
from botbuilder.schema import ActionTypes, Activity, ActivityTypes
from botbuilder.core import MessageFactory, CardFactory, TurnContext
from botbuilder.schema import ActionTypes, HeroCard, CardAction, Attachment

class BotFormattingCards(TeamsActivityHandler):
    def __init__(self):
        # Constructor for the TeamsBot class. Initializes the parent class (TeamsActivityHandler)
        super().__init__()

    async def on_teams_members_added(
        self,
        teams_members_added: list[TeamsChannelAccount],
        team_info: TeamInfo,
        turn_context: TurnContext,
    ):

        await self.send_welcome_message(turn_context)

    async def on_message_activity(self, turn_context: TurnContext):
        text = turn_context.activity.text

        # List of valid card options
        adaptive_format_cards = [
            "CodeBlock",
            "MentionSupport",
            "InfoMasking",
            "FullWidthCard",
            "StageViewImages",
            "OverflowMenu",
            "HTMLConnector",
            "CardWithEmoji",
            "Persona",
            "PersonaSet",
            "Layout",
            "Borders",
            "RoundedCorners",
            "FluentIcons",
            "MediaElements",
            "StarRatings",
            "ConditionalCard",
            "ScrollableContainer",
            "CompoundButton",
            "ContainerLayout",
            "DonutChart",
            "GaugeChart",
            "HorizontalChart",
            "HorizontalChartStacked",
            "LineChart",
            "PieChart",
            "VerticalBarChart",
            "VerticalBarGroupedChart",
        ]

        # Check if the received text is a valid card option
        if text in adaptive_format_cards:
            card_method_map = {
                "MentionSupport": cards.mentionSupport(),
                "InfoMasking": cards.informationMaskingCard(),
                "FullWidthCard": cards.sampleAdaptiveWithFullWidth(),
                "StageViewImages": cards.stageViewForImages(),
                "OverflowMenu": cards.overflowMenu(),
                "HTMLConnector": cards.formatHTMLConnectorCard(),
                "CardWithEmoji": cards.adaptiveCardWithEmoji(),
                "Persona": cards.adaptivePeoplePersonaCardIcon(),
                "PersonaSet": cards.adaptivePeoplePersonaCardSetIcon(),
                "CodeBlock": cards.codeBlocksCard(),
                "Layout": cards.AdaptiveCardResponsiveLayout(),
                "Borders": cards.adaptiveCardBorders(),
                "RoundedCorners": cards.adaptiveCardRoundedCorners(),
                "FluentIcons": cards.adaptiveCardFluentIcon(),
                "MediaElements": cards.adaptiveCardMediaElements(),
                "StarRatings": cards.adaptiveCardStarRatings(),
                "ConditionalCard": cards.adaptiveCardConditional(),
                "ScrollableContainer": cards.adaptiveCardScrollable(),
                "CompoundButton": cards.adaptiveCardCompoundButton(),
                "ContainerLayout": cards.adaptiveCardContainerLayouts(),
                "DonutChart": cards.adaptiveCardDonutChart(),
                "GaugeChart": cards.adaptiveCardGaugeChart(),
                "HorizontalChart": cards.adaptiveCardHorizontalBarChart(),
                "HorizontalChartStacked": cards.adaptiveCardHorizontalBarStacked(),
                "LineChart": cards.adaptiveCardLineChart(),
                "PieChart": cards.adaptiveCardPieChart(),
                "VerticalBarChart": cards.adaptiveCardVerticalBarChart(),
                "VerticalBarGroupedChart": cards.adaptiveCardVerticalBarGroupedChart(),
            }

        # Call the respective method dynamically
        if text in card_method_map:
            # Get the attachment directly (it is already an object, not a function)
            card_attachment = card_method_map[text]

            # Ensure the attachment is a list as expected
            if isinstance(card_attachment, list):
                attachments = card_attachment
            else:
                attachments = [card_attachment]

            # For only HTML Connector Card
            if text == "HTMLConnector":
                attachment = Attachment(
                    content_type="application/vnd.microsoft.teams.card.o365connector",
                    content=cards.formatHTMLConnectorCard(),
                )
                await turn_context.send_activity(
                    Activity(type=ActivityTypes.message, attachments=[attachment])
                )
            else:
                await turn_context.send_activity(
                    Activity(type=ActivityTypes.message, attachments=attachments)
                )

            await turn_context.send_activity(f"You have selected <b>{text}</b>")

        elif (
            turn_context.activity.value is not None
            and turn_context.activity.text is None
        ):
            activity_value = turn_context.activity.value

            # Check if the activity contains star rating feedback
            if "rating1" in activity_value and "rating2" in activity_value:
                await turn_context.send_activity(f"Ratings Feedback: {activity_value}")

        # Send adaptive card formats
        await self.send_adaptive_card_formats(turn_context)

    async def send_welcome_message(self, turn_context: TurnContext):
        activity = turn_context.activity

        # Iterate over all new members added to the conversation
        for member in activity.members_added:
            if member.id != activity.recipient.id:
                welcome_message = (
                    "Welcome to Adaptive Card Format. This bot will introduce you to different types of formats. "
                    "Please select a card from the options provided."
                )

            await turn_context.send_activity(welcome_message)

            # Send the adaptive card formats
            await self.send_adaptive_card_formats(turn_context)

    def send_mention_support_card(self):
        return cards.mentionSupport()

    def send_info_masking_card(self):
        return cards.informationMaskingCard()

    def send_fullwidthcard_card(self):
        return cards.sampleAdaptiveWithFullWidth()

    async def send_adaptive_card_formats(self, turn_context: TurnContext):
        card_actions = [
            CardAction(
                type=ActionTypes.im_back, title="MentionSupport", value="MentionSupport"
            ),
            CardAction(
                type=ActionTypes.im_back, title="InfoMasking", value="InfoMasking"
            ),
            CardAction(
                type=ActionTypes.im_back, title="FullWidthCard", value="FullWidthCard"
            ),
            CardAction(
                type=ActionTypes.im_back,
                title="StageViewImages",
                value="StageViewImages",
            ),
            CardAction(
                type=ActionTypes.im_back, title="OverflowMenu", value="OverflowMenu"
            ),
            CardAction(
                type=ActionTypes.im_back, title="HTMLConnector", value="HTMLConnector"
            ),
            CardAction(
                type=ActionTypes.im_back, title="CardWithEmoji", value="CardWithEmoji"
            ),
            CardAction(type=ActionTypes.im_back, title="Persona", value="Persona"),
            CardAction(
                type=ActionTypes.im_back, title="PersonaSet", value="PersonaSet"
            ),
            CardAction(type=ActionTypes.im_back, title="CodeBlock", value="CodeBlock"),
            CardAction(type=ActionTypes.im_back, title="Layout", value="Layout"),
            CardAction(type=ActionTypes.im_back, title="Borders", value="Borders"),
            CardAction(
                type=ActionTypes.im_back, title="RoundedCorners", value="RoundedCorners"
            ),
            CardAction(
                type=ActionTypes.im_back, title="FluentIcons", value="FluentIcons"
            ),
            CardAction(
                type=ActionTypes.im_back, title="MediaElements", value="MediaElements"
            ),
            CardAction(
                type=ActionTypes.im_back, title="StarRatings", value="StarRatings"
            ),
            CardAction(
                type=ActionTypes.im_back,
                title="ConditionalCard",
                value="ConditionalCard",
            ),
            CardAction(
                type=ActionTypes.im_back,
                title="ScrollableContainer",
                value="ScrollableContainer",
            ),
            CardAction(
                type=ActionTypes.im_back, title="CompoundButton", value="CompoundButton"
            ),
            CardAction(
                type=ActionTypes.im_back,
                title="ContainerLayout",
                value="ContainerLayout",
            ),
            CardAction(
                type=ActionTypes.im_back, title="DonutChart", value="DonutChart"
            ),
            CardAction(
                type=ActionTypes.im_back, title="GaugeChart", value="GaugeChart"
            ),
            CardAction(
                type=ActionTypes.im_back,
                title="HorizontalChart",
                value="HorizontalChart",
            ),
            CardAction(
                type=ActionTypes.im_back,
                title="HorizontalChartStacked",
                value="HorizontalChartStacked",
            ),
            CardAction(type=ActionTypes.im_back, title="LineChart", value="LineChart"),
            CardAction(type=ActionTypes.im_back, title="PieChart", value="PieChart"),
            CardAction(
                type=ActionTypes.im_back,
                title="VerticalBarChart",
                value="VerticalBarChart",
            ),
            CardAction(
                type=ActionTypes.im_back,
                title="VerticalBarGroupedChart",
                value="VerticalBarGroupedChart",
            ),
        ]

        await self.send_welcome_card(turn_context, card_actions)

    async def send_welcome_card(self, context: TurnContext, card_actions):
        card = HeroCard(
            title="Please select a card from given options.", buttons=card_actions
        )

        attachment = CardFactory.hero_card(card)
        await context.send_activity(MessageFactory.attachment(attachment))
